ads/google/a4a/test/test-performance.js

```diff
@@ -0,0 +1,307 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpAdLifecycleReporter} from '../performance';
+import {EXPERIMENT_ATTRIBUTE} from '../traffic-experiments';
+import {childElements} from '../../../../src/dom';
+import {createIframePromise} from '../../../../testing/iframe';
+import * as sinon from 'sinon';
+
+/**
+ * Verify that `address` matches all of the patterns in `matchlist`.
+ *
+ * @param {!string} address
+ * @param {!Array<!RegExp>} matchList
+ */
+function expectMatchesAll(address, matchList) {
+  matchList.forEach(m => {
+    expect(address).to.match(m);
+  });
+}
+
+/**
+ * Whether `element` is an `img` DOM node.
+ * @param {!Element} element
+ * @returns {boolean}
+ */
+function isImgNode(element) {
+  return element.tagName == 'IMG';
+}
+
+/**
+ * Verify that `element` has at least one sibling DOM node that is an
+ * `img` tag whose `src` matches all of the patterns in `matchList`.
+ *
+ * @param {!Element} element
+ * @param {!Array<!RegExp>} matchList
+ */
+function expectHasSiblingImgMatchingAll(element, matchList) {
+  const imgSiblings = childElements(
+      element.parentElement, e => isImgNode(e));
+  expect(imgSiblings).to.not.be.empty;
+  const result = imgSiblings.some(e => {
+    const src = e.getAttribute('src');
+    return matchList.map(m => m.test(src)).every(x => x);
+  });
+  expect(result, 'No element sibling of ' + element + ' matched all patterns')
+      .to.be.true;
+}
+
+// From
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions?redirectlocale=en-US&redirectslug=JavaScript%2FGuide%2FRegular_Expressions
+/**
+ * Escape a string for use as a literal inside a regular expression.
+ * @param {!string} string
+ * @returns {!string}
+ * @private
+ */
+function escapeRegExp_(string) {
+  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
+}
+
+describe('AmpAdLifecycleReporter', () => {
+  let sandbox;
+  let emitPingSpy;
+  beforeEach(() => {
+    sandbox = sinon.sandbox.create();
+    emitPingSpy = sandbox.spy(AmpAdLifecycleReporter.prototype, 'emitPing_');
+  });
+  afterEach(() => {
+    sandbox.restore();
+  });
+
+  describe('#sendPing', () => {
+    it('should request a single ping and insert into DOM', () => {
+      return createIframePromise(false).then(fixture => {
+        const win = fixture.win;
+        const doc = fixture.doc;
+        const elem = doc.createElement('div');
+        doc.body.appendChild(elem);
+        const reporter = new AmpAdLifecycleReporter(win, elem, 'test_foo');
```

Comment:
probably it's worth to extract the fixture logic into `beforeEach`. you can do return promise in `beforeEach`